erts/emulator/beam/erl_bits.c

```diff
@@ -2261,18 +2269,140 @@ int erts_cmp_bits__(const byte *a_ptr,
+    doffs = BIT_OFFSET(doffs);
+    deoffs = BIT_OFFSET(doffs+n);
+    lmask = (doffs) ? MAKE_MASK(8-doffs) : 0;
+    rmask = (deoffs) ? (MAKE_MASK(deoffs)<<(8-deoffs)) : 0;
 
-void 
-erts_copy_bits(const byte* src, /* Base pointer to source. */
+    /*
+     * Take care of the case that all bits are in the same byte.
+     */
+
+    if (doffs+n < 8) {		/* All bits are in the same byte */
+        lmask = (lmask & rmask) ? (lmask & rmask) : (lmask | rmask);
+
+        if (soffs == doffs) {
+            *dst = MASK_BITS(*src, *dst, lmask);
+        } else if (soffs > doffs) {
+            Uint bits = (*src << (soffs-doffs));
+            if (soffs+n > 8) {
+                src++;
+                bits |= (*src >> (8-(soffs-doffs)));
+            }
+            *dst = MASK_BITS(bits, *dst, lmask);
+        } else {
+            *dst = MASK_BITS((*src >> (doffs-soffs)), *dst, lmask);
+        }
+        return;			/* We are done! */
+    }
+
+    /*
+     * At this point, we know that the bits are in 2 or more bytes.
+     */
+
+    count = (lmask ? (n - (8 - doffs)) : n) >> 3;
+
+    if (soffs == doffs) {
+        /*
+         * The bits are aligned in the same way. We can just copy the bytes
+         * (except for the first and last bytes).
+         */
+
+        if (lmask) {
+            *dst = MASK_BITS(*src, *dst, lmask);
+            dst++, src++;
+        }
+
+        sys_memcpy(dst, src, count);
+
+        if (rmask) {
+            dst += count;
+            src += count;
+            *dst = MASK_BITS(*src, *dst, rmask);
+        }
+    } else {
+        Uint bits;
+        Uint bits1;
+        Uint rshift;
+        Uint lshift;
+
+        /*
+         * The tricky case. The bits must be shifted into position.
+         */
+
+        if (soffs > doffs) {
+            lshift = soffs - doffs;
+            rshift = 8 - lshift;
+            bits = *src;
+            if (soffs + n > 8) {
+                src++;
+            }
+        } else {
+            rshift = doffs - soffs;
+            lshift = 8 - rshift;
+            bits = 0;
+        }
+
+        if (lmask) {
+            bits1 = bits << lshift;
+            bits = *src++;
+            bits1 |= (bits >> rshift);
+            *dst = MASK_BITS(bits1, *dst, lmask);
+            dst++;
+        }
+
+        while (count--) {
+            bits1 = bits << lshift;
+            bits = *src++;
+            *dst = bits1 | (bits >> rshift);
+            dst++;
+        }
+
+        if (rmask) {
+            bits1 = bits << lshift;
+            if ((rmask << rshift) & 0xff) {
+                bits = *src;
+                bits1 |= (bits >> rshift);
+            }
+            *dst = MASK_BITS(bits1, *dst, rmask);
+        }
+    }
+}
+
+/*
+ * The basic bit copy operation. Copies n bits from the source buffer to
+ * the destination buffer. The source pointer is incremented, whereas the
+ * destination pointer is decremented, resulting in a reversal of the
+ * stored bytes.
```

Comment:
s/bytes/bits/ ?

Reply:
Actually not. It does not reverse bit order, only byte order.

I've pushed a fixup commit in which I try to explain it more clearly.

Reply:
Much clearer. I was wondering about what would happen if the bits weren't byte aligned, with the new wording that also becomes clear.